crypto/pkcs7/pk7_smime.c

```diff
@@ -320,8 +320,12 @@ int PKCS7_verify(PKCS7 *p7, STACK_OF(X509) *certs, X509_STORE *store,
         tmpout = out;
 
     /* We now have to 'read' from p7bio to calculate digests etc. */
-    if ((buf = OPENSSL_malloc(BUFFERSIZE)) == NULL)
+    if ((buf = OPENSSL_malloc(BUFFERSIZE)) == NULL) {
+        if (flags & PKCS7_TEXT)
```

Comment:
Its probably cleaner to move these 2 lines to the err block and then just remove all other BIO_free(tmpout); calls

tmpout is set to NULL initially so this should work regardless of the code path.

Reply:
Looks good. I have submitted a new commit to make it cleaner.